test/others/criu-ns/run.py

```diff
@@ -0,0 +1,181 @@
+    return ret
+
+
+def criu_ns_restore(shell_job=False, restore_detached=False):
+    if shell_job:
+        cmd = [CRIU_NS, "restore", "-D", "dumpdir", "-v4",
+               "--shell-job", "--criu-binary", CRIU_BIN]
+    else:
+        cmd = [CRIU_NS, "restore", "-D", "dumpdir", "-v4",
+               "--criu-binary", CRIU_BIN]
+    if restore_detached:
+        cmd = [CRIU_NS, "restore", "-D", "dumpdir", "-v4",
+               "--restore-detached", "--pidfile", PIDFILE,
+               "--criu-binary", CRIU_BIN]
+    ret = subprocess.Popen(cmd).wait()
+    return ret
+
+
+def test_dump_and_restore_with_shell_job():
+    """Test criu-ns dump and restore with --shell-job option"""
+    check_dumpdir()
+    os.setsid()
+
+    with open("running", "w") as file:
+        file.name
+    pid = os.fork()
+    if pid == 0:
+        while True:
+            if not os.access("running", os.F_OK):
+                sys.exit(0)
+            time.sleep(1)
+
+    ret = criu_ns_dump(pid, shell_job=True)
+    if ret != 0:
+        sys.exit(ret)
+
+    os.unlink("running")
+    fd_m, fd_s = create_pty()
+    pid = os.fork()
+    if pid == 0:
+        os.setsid()
+        fd_m.close()
+        # since criu-ns takes control of the tty stdin
+        os.dup2(fd_s.fileno(), 0)
+        ret = criu_ns_restore(shell_job=True)
+        if ret != 0:
+            sys.exit(ret)
+        os._exit(0)
+
+    fd_s.close()
+    os.waitpid(pid, 0)
+
+
+def test_dump_and_restore_without_shell_job(restore_detached=False):
+    """Test criu-ns dump and restore with an isolated process"""
+    check_dumpdir()
+    pid, fd_s = create_isolated_dumpee()
+    ret = criu_ns_dump(pid)
+    if ret != 0:
+        sys.exit(ret)
+
+    if not restore_detached:
+        os.unlink("running")
+
+    fd_m, fd_s = create_pty()
+    pid = os.fork()
+    if pid == 0:
+        os.setsid()
+        ret = criu_ns_restore(restore_detached=restore_detached)
+        if ret != 0:
+            sys.exit(ret)
+        os._exit(0)
+
+    fd_m.close()
+    fd_s.close()
+    os.waitpid(pid, 0)
+
+
+def test_dump_and_restore_in_pidns():
+    """Test criu-ns dump and restore in namespaces"""
+    def _dump():
+        pid, fd_s = create_isolated_dumpee()
+        ret = criu_ns_dump(pid)
+        if ret != 0:
+            sys.exit(ret)
+
+    def _restore():
+        ret = criu_ns_restore(restore_detached=True)
+        if ret != 0:
+            sys.exit(ret)
+
+    def _get_restored_pid():
+        restored_pid = 0
+        pidfile_path = os.path.join(IMG_DIR, PIDFILE)
+        if not os.path.exists(pidfile_path):
+            raise FileNotFoundError("pidfile not found")
+        with open(pidfile_path, "r") as pidfile:
+            restored_pid = pidfile.read().strip()
+        return int(restored_pid)
+
+    def _redump():
+        try:
+            restored_pid = _get_restored_pid()
+        except FileNotFoundError:
+            sys.exit(1)
+        ret = criu_ns_dump(restored_pid)
```

Comment:
Before doing second dump we can `IMG_DIR = "dumpdir2", so that old image files don't mess up with the new ones. Please do.

Reply:
done.